internal/resolver/delegatingresolver/delegatingresolver_test.go

```diff
@@ -0,0 +1,319 @@
+)
+
+type s struct {
+	grpctest.Tester
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+const (
+	targetTestAddr          = "test.com"
+	resolvedTargetTestAddr  = "1.2.3.4:8080"
+	resolvedTargetTestAddr1 = "1.2.3.5:8080"
+	envProxyAddr            = "proxytest.com"
+	resolvedProxyTestAddr   = "2.3.4.5:7687"
+	resolvedProxyTestAddr1  = "2.3.4.6:7687"
+)
+
+// overwriteAndRestore overwrite function HTTPSProxyFromEnvironment and
+// returns a function to restore the default values.
+func overwrite(hpfe func(req *http.Request) (*url.URL, error)) func() {
+	backHPFE := HTTPSProxyFromEnvironment
+	HTTPSProxyFromEnvironment = hpfe
+	return func() {
+		HTTPSProxyFromEnvironment = backHPFE
+	}
+}
+
+func (s) TestUpdateProxyUrlEnv(t *testing.T) {
+	// Overwrite the function in the test and restore them in defer.
+	hpfe := func(req *http.Request) (*url.URL, error) {
+		if req.URL.Host == targetTestAddr {
+			return &url.URL{
+				Scheme: "https",
+				Host:   envProxyAddr,
+			}, nil
+		}
+		return nil, nil
+	}
+	defer overwrite(hpfe)()
+
+	// envTestAddr should be handled by ProxyFromEnvironment.
+	got, err := updateProxyURL(targetTestAddr)
+	if err != nil {
+		t.Error(err)
+	}
+	if got.Host != envProxyAddr {
+		t.Errorf("want %v, got %v", envProxyAddr, got)
+	}
+}
+
+// createTestResolverClientConn initializes a test ResolverClientConn and
+// returns it along with channels for resolver state updates and errors.
+func createTestResolverClientConn(t *testing.T) (*testutils.ResolverClientConn, chan resolver.State, chan error) {
+	stateCh := make(chan resolver.State, 1)
+	errCh := make(chan error, 1)
+
+	tcc := &testutils.ResolverClientConn{
+		Logger:       t,
+		UpdateStateF: func(s resolver.State) error { stateCh <- s; return nil },
+		ReportErrorF: func(err error) { errCh <- err },
+	}
+	return tcc, stateCh, errCh
+}
+
+// overwriteAndRestoreProxyEnv overwrites the proxy environment variable and
+// returns a function to restore the default values.
+func overwriteAndRestoreProxyEnv(proxyURI string) func() {
+	origHTTPSProxy := envconfig.HTTPSProxy
+	envconfig.HTTPSProxy = proxyURI
+	return func() { envconfig.HTTPSProxy = origHTTPSProxy }
+}
+
+// TestDelegatingResolverNoProxy verifies the behavior of the delegating resolver
+// when no proxy is configured.
+func (s) TestDelegatingResolverNoProxy(t *testing.T) {
+	// Enable HTTP Proxy env var.
+	defer overwriteAndRestoreProxyEnv("")()
+	mr := manual.NewBuilderWithScheme("test") // Set up a manual resolver to control the address resolution.
+	target := "test:///" + targetTestAddr
+
+	tcc, stateCh, _ := createTestResolverClientConn(t)
+	// Create a delegating resolver with no proxy configuration
+	dr, err := New(resolver.Target{URL: *testutils.MustParseURL(target)}, tcc, resolver.BuildOptions{}, mr, false)
+	if err != nil || dr == nil {
+		t.Fatalf("Failed to create delegating resolver: %v", err)
+	}
+
+	// Update the manual resolver with a test address.
+	mr.UpdateState(resolver.State{
```

Comment:
shouldn't we check before updating state as will that delegating resolver's target resolver got the correct addresses? it is because the `New` does the `resolver.build()` as well in case of no proxy

Reply:
In the no proxy case, we directly pass the `ClientConn` to the targetResolver instead of the `wrappingClientConn` inplemented by delegating resolver. So for the no proxy case, the delegating resolver will not get the address, but will rather be passed to the `ClientConn` directly.